### DISCUSSION TITLE
Airflow 2.7.1 Webserver Errors for CA Cert

### DISCUSSION THREAD

**Discussion by tagoreghanta** [apache/airflow#34780] [Q&A] [answered]
Hi All,

I have airflow 2.7.1, trying to enable for https  . I could do with SelfSigned cert 

web_server_ssl_cert = /tmp/airflowtest.pem
web_server_ssl_key = /tmp/privatekey.key

But when I am trying to enable SSL with a DigiCert  with a  Pass Phrase keeps erroring 

web_server_ssl_cert = /tmp/airflowtest.pem  ( This is Combined PEM including Intermediate and Root Cert )
web_server_ssl_key = /tmp/privatekey.key    

access the airflow through browser , application errors
===================================================
Enter PEM pass phrase:
[2023-10-05 13:57:42 +0000] [2071147] [INFO] Handling signal: ttou
[2023-10-05 13:57:42 +0000] [2071182] [INFO] Worker exiting (pid: 2071182)
[2023-10-05 13:57:42 +0000] [2071185] [ERROR] Socket error processing request.
Traceback (most recent call last):
  File "/home/ubuntu/venv/lib/python3.10/site-packages/gunicorn/workers/sync.py", line 132, in handle
    client = sock.ssl_wrap_socket(client, self.cfg)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/gunicorn/sock.py", line 229, in ssl_wrap_socket
    return ssl_context(conf).wrap_socket(sock,
  File "/home/ubuntu/venv/lib/python3.10/site-packages/gunicorn/sock.py", line 225, in ssl_context
    return conf.ssl_context(conf, default_ssl_context_factory)
  File "/home/ubuntu/venv/lib/python3.10/site-packages/gunicorn/config.py", line 2005, in ssl_context
    return default_ssl_context_factory()
  File "/home/ubuntu/venv/lib/python3.10/site-packages/gunicorn/sock.py", line 219, in default_ssl_context_factory
    context.load_cert_chain(certfile=conf.certfile, keyfile=conf.keyfile)
OSError: [Errno 22] Invalid argument

==============
I do not see in airflow.cfg documentation  on how to pass "PEM pass phrase"

**Comment by Taragolis** [comment, marked as answer]
The implementation of certificates based on Flask implementation, so it basically limits on this implementation.

IMO better place Airflow webserver behind reverse proxy, and configure HTTPs traffic there
- regular http servers could do that [nginx](https://www.nginx.com/), [Apache HTTP Server,](https://httpd.apache.org/), etc
- managed application load balancer e.g. AWS ALB or other cloud providers balancers

There is instruction how to configure nginx as reverse proxy: [Running Airflow behind a reverse proxy](https://airflow.apache.org/docs/apache-airflow/stable/howto/run-behind-proxy.html)

**Comment by tagoreghanta** [comment]
Thank you for your suggestion , I could get past my issue modifying 
venv/lib/python3.10/site-packages/gunicorn/sock.py     
context.load_cert_chain(certfile=conf.certfile, keyfile=conf.keyfile, password="password")

Planning to pass it as a variable

I am running into subsequent issue which I am trying to fix though 


 {views.py:660} ERROR - Error authorizing OAuth access token: HTTPSConnectionPool(host='login.microsoftonline.com', port=443): Max retries exceeded with url: /..../oauth2/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))
 {views.py:660} ERROR - Error authorizing OAuth access token: HTTPSConnectionPool(host='login.microsoftonline.com', port=443): Max retries exceeded with url: /2.....f/oauth2/token (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate in certificate chain (_ssl.c:1007)')))
 "GET /oauth-authorized/azure?code=....... HTTP/1.1" 302 201 "https://airflowtest.......:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36"
  "GET /login/ HTTP/1.1" 200 19439 "https://airflowtest.......com:8080/" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36"


Looks like OAUTH side issue I am triaging

**Reply by Taragolis** [reply]
Seems like error happen due to self-signed certificate. So you need use certificate from one of [trusted authorities](https://en.wikipedia.org/wiki/Certificate_authority) or allow to use self-signed certificates (maybe you could find the answer in [Security: Webserver](https://airflow.apache.org/docs/apache-airflow/stable/security/webserver.html))
